_posts/2017-08-18-Refactoring-with-reactJS-shared-state.md

```diff
@@ -0,0 +1,252 @@
+
+# The Problem
+
+Let's look at the `FailureBanner` component from last time.
+
+### Original FailureBanner.jsx
+{% highlight javascript linenos %}
+import React from 'react';
+
+export default class FailureBanner extends React.Component {
+    render(){
+        if(this.props.isShown) {
+            // displays an alert with given message property when isShown is true
+            return (
+                <div className="flamingo">
+                    <div role="alert" className="alert alert-danger" id="msg-text">
+                        <div className="alert-left flex-center">
+                            <i className="fa fa-ban fa-lg" aria-hidden="true"></i>
+                        </div>
+                        <p>{this.props.message}</p>
+                    </div>
+                </div>
+            );
+        }
+        // when hidden, render nothing
+        return null;
+    }
+}
+{% endhighlight %}
+
+Using our `FailureBanner` example from last time, let's suppose that we want to display the failure banner to a user only after they've clicked a button on the page. Let's assume that the button is rendered by the server and just exists on the existing page.
+
+{% highlight html linenos %}
+<script type="text/javascript">
+    $('#failureButton').click(function(e) {
+        // LEGACY CODE - does some crazy stuff to the rest of the page
+
+        // Here is where we want to display our React failure banner with a specific message
+    });
+</script>
+
+<div data-react-component='FailureBanner' data-is-shown='false' data-message=''></div>
+<button id='failureButton' >Do The Thing</button>
+
+<!-- Include the react initialization code from last time
+It renders the FailureBanner on page load with the isShown and message props set to the values in the tag -->
+<script src='index.js'></script>
+{% endhighlight %}
+
+In this example, we already have an instance of the `FailureBanner` rendered into the page with some default properties. In this case, it's hidden from the user because we set the isShown prop to `false`. Once the user clicks our legacy failure button, the application does a bunch of stuff. At the very end, we've decided that it should now show a failure banner to the user. We want to update the `FailureBanner` component's props so that `isShown` is true and the `message` says "Sorry, we couldn't do the thing."
+
+# How do we update component props?
+
+One solution we’ve seen before is to unmount the React component, then re-render the component into the DOM with the updated props. It’s a workable solution, but we’ve always disliked it because it completely ignores one of the biggest benefits of using React: the [Virtual DOM](https://www.codecademy.com/articles/react-virtual-dom).
+
```

Comment:
Trying to keep our proof of concept to a minimum we have attempted to capture components at the time of creation and inject some form of function or object to give us global access to individual component's state. After several attempts with marginal degree of success we realized that we are creating a sufficiently complex  home grown system that attempts to replicate existing technologies. For the sake of brevity and community support  we decided to turned to industry accepted technologies already used elsewhere in-house.